[R1] **POSITIVE**

Reviewed at `eb73358` against base `d320ca13`, in a throwaway clone under `/tmp`. Every claim in the PR body checks out, including the mutation numbers, and I verified claim 4 against the CI-pinned tsn-gen rev rather than the local working tree. The code change is correct and its blast radius really is zero.

The self-test is real, not decorative: it catches the reverted grader and three of the four wrong-but-plausible mutations I threw at it. It does have four survivors, one of them sharp, and I report those below. None of them makes the tree worse than it is today, so they are asks and not blockers, and two of them are one line each.

---

## 1. The premise holds, and by construction

`git show d320ca13:tb/verilator/tsn_fuzz/fuzz_ptp.py | sed -n '253,282p'` is a bare `if "value" / elif "range" / elif "values"` chain with no `else`. The only other exit is the `continue` at the top of the loop body, which is gated on `name in skip or name not in got or not con`. So a field with a non-empty `con` that is none of the three kinds executes nothing at all: no `rep.ck`, no `rep.gap`, no `print`, no `continue` line. The check does not exist.

Confirmed empirically, which is the part that makes it undeniable. Reverted grader, self-test output:

```
  [FAIL] mask: no undefined bit set                     pass=0/1 fail=0/0
  [FAIL] mask: every defined bit set is still legal     pass=0/1 fail=0/0
```

`pass=0/1 fail=0/0`. Neither a pass nor a failure. That is the silent skip on the record.

## 2. The mask semantics are right, and the agreement is wider than the PR claims

The PR argues agreement with `tsn_model.py`. That holds:

* `legal()` yields `[0, m] + [1 << b for b in range(bits) if m & (1 << b)]`, every one of which satisfies `v & ~m == 0`;
* `illegal()` yields `[undef] + single undefined bits`, where `undef = (~m) & full`, every one of which fails it.

So each generated legal probe passes the new check and each generated illegal probe fails it. Exact.

But `tsn_model.py` is this repo's re-reading of the model, so I went to the generator itself, at the pinned rev:

```
$ git show fde6520:docs/03_protocol_yaml_reference.md | grep -n mask
67:| `expected.mask` | No | `[M]` | Only bits within `M` may be set in the random value. |
97:      mask: [0xC0000000]       # only the top two bits may be set

$ git show fde6520:traffic-gen/src/packet_builder.cpp
    /* Bitmask: only bits within the mask may be set. */
    if (var.hasMask()) {
        value &= var.getMask();
    }
```

tsn-gen's reference doc, its authoring guide, and its C++ builder all say the same thing the new branch asserts. Grader, generator and builder agree. Not a blocker, the opposite: this is the strongest-grounded part of the change.

Spot-checked two of the seven live masks against the standard, and the model reading is right there too: `acquire_entity_flags` mask `0xC0000000` is exactly PERSISTENT plus RELEASE, and `lock_entity_flags` mask `0x80000000` is exactly UNLOCK.

## 3. The mutation claim reproduces exactly

```
$ python3 test_grade_tx.py            # PR as merged
grade_tx self-test: 12 checks: 12 PASS, 0 FAIL      rc=0

$ # grade_tx reverted to its pre-#146 shape
$ python3 test_grade_tx.py
  [FAIL] mask: no undefined bit set                     pass=0/1 fail=0/0
  [FAIL] mask: an undefined bit IS set                  pass=0/0 fail=0/1  needle='undefined=0x80' MISSING in ''
  [FAIL] mask: every defined bit set is still legal     pass=0/1 fail=0/0
  [FAIL] unknown kind fails closed                      pass=0/0 fail=0/1  needle='ungradeable constraint kind' MISSING in ''
  [FAIL] unknown kind names the field                   pass=0/0 fail=0/1  needle='t.f' MISSING in ''
grade_tx self-test: 12 checks: 7 PASS, 5 FAIL       rc=1

$ make graderself                                    MAKE_RC=2
```

7 PASS, 5 FAIL, exit 1, `make graderself` exit 2, and the five are the five the PR names. Verbatim.

## 4. No tally movement, verified against the PIN

Read straight out of `git show fde652067d3618f47fba2c2e8eb05f6c9f6fa97f:<path>`, never the working tree. Constraint-kind histogram over the whole `protocols/` tree at that rev:

```
  8021as_announce.yaml               {'value': 16, 'values': 1, 'range': 1, '<none>': 10}
  8021as_eth_header.yaml             {'value': 2, '<none>': 1}
  8021as_follow_up.yaml              {'value': 14, 'values': 1, 'range': 2, '<none>': 9}
  8021as_pdelay_req.yaml             {'value': 13, 'range': 1, '<none>': 3}
  8021as_pdelay_resp.yaml            {'value': 11, 'range': 3, '<none>': 5}
  8021as_pdelay_resp_follow_up.yaml  {'value': 11, 'range': 3, '<none>': 5}
  8021as_sync.yaml                   {'value': 12, 'values': 1, 'range': 1, '<none>': 2}

  TOTAL across all protocols/: {'value': 135, 'values': 67, 'range': 29, 'mask': 7}
```

Zero `mask` and zero unrecognised kinds in all seven PTP models, so the new branches are unreachable from `fuzz_ptp`. The seven live masks are exactly the seven the PR names, in exactly the five files it names. No field anywhere in the tree declares two constraint kinds.

Rather than stop at exhaustion over the model set, I drove the real pinned models through both graders:

```
### AFTER (PR #147)                      ### BEFORE (pre-#146)
  8021as_announce.yaml         15          8021as_announce.yaml         15
  8021as_eth_header.yaml        2          8021as_eth_header.yaml        2
  8021as_follow_up.yaml        17          8021as_follow_up.yaml        17
  8021as_pdelay_req.yaml       14          8021as_pdelay_req.yaml       14
  8021as_pdelay_resp.yaml      14          8021as_pdelay_resp.yaml      14
  8021as_pdelay_resp_fu.yaml   14          8021as_pdelay_resp_fu.yaml   14
  8021as_sync.yaml             14          8021as_sync.yaml             14
  TOTAL: 90                                TOTAL: 90
```

Identical. Zero checks added, zero removed. Claim confirmed.

For the record on the trap: local `tsn-gen` HEAD `f52f1f9a` is NOT a descendant of the pin, but `git diff fde6520 HEAD -- protocols/data_link/ptp/` is empty, so the working tree happens to give the same answer here. That was luck, not method.

## 5. The deliberate exclusion is the right call, but the deferral is not backed

Failing on an absent `expected:` would be wrong, and I agree with the PR. The two cases are different claims: an unknown kind means the model asserts something the grader cannot read, while an absent block means the model asserts nothing. Only the first should fail closed. Concretely, at the pin the seven PTP models carry 35 unconstrained fields, so the alternative reddens 35 legitimate fields.

What is not backed is "is tracked separately". `gh issue list --state all --limit 100` plus two `gh api search/issues` passes over titles and bodies return no such issue. The only issues mentioning it are #146, which this PR closes, and #147 itself. #148 and #149 are the stale gPTP TEST_RESULTS work, unrelated.

That matters more than usual here, because deleting a pin is how the campaign already shrank once without anyone noticing, and #146's own text waves it off with "That one is visible in the tally at least". Nothing gates the tally against a previous run, so it is visible only to someone who happens to compare. **Please file the issue before merging this one**, otherwise closing #146 closes the file on both halves of the hole.

## 6. Output hygiene is clean

The `redirect_stdout` cannot hide a real failure, because the verdict never reads stdout: `grade()` returns `rep.npass`, `rep.nfail` and `rep.failures[0]`, which are object state that `Report.ck` updates before it prints. An exception inside the `with` escapes the context manager, prints to stderr, and exits non-zero, so that path is loud too.

`suite_tally.py` is not confused. Scanning the real graderself output:

```
  checks=12 failures=0
  matched  = [('n-checks-pass-fail', 12, 0, 'grade_tx self-test: 12 checks: 12 PASS, 0 FAIL')]
  UNPARSED = []
```

One clean match, no UNPARSED. The twelve per-case lines (`pass=1/1 fail=0/0`) trip neither `SHAPES` nor `TALLY_SHAPED`, since the latter wants `N pass,` with a space and a comma. `--campaign-guard` reads only `obj_aaf/aaf.out` and `obj_ptp/ptp.out`, which `graderself` never writes, so the two campaign guards are untouched.

Worth stating out loud so nobody is surprised: the tsn_fuzz **suite** headline gains +12 checks, because `run_all_suites.sh` redirects all of `make` into the suite log and `suite_tally` reads that line. That is honest new work and not a contradiction of the PR, whose "no tally movement" claim is about the campaign. Just flagging it so the next sweep delta is expected.

## 7. Makefile placement is correct

With an emptied environment, no verilator on PATH and a bogus generator root:

```
$ env -i PATH=/usr/bin:/bin HOME=/nonexistent TSN_GEN_ROOT=/nonexistent/tsn-gen \
        VERILATOR=/nonexistent/verilator make graderself
grade_tx self-test: 12 checks: 12 PASS, 0 FAIL      RC=0
```

Needs neither Verilator nor tsn-gen, as claimed. And a failing self-test does abort the whole run, before `build`:

```
$ env -i ... make run
grade_tx self-test: 12 checks: 7 PASS, 5 FAIL
make: *** [Makefile:96: graderself] Error 1              MAKE_RUN_RC=2
```

`run_all_suites.sh:168` invokes `timeout "$TMO" make -C "$d"` with no `-j`, so prerequisites stay serial and `graderself` really is first.

---

## Findings: what the self-test does NOT catch

The PR proves the forward direction. I ran the converse. Seven mutations, four survive with the self-test at **12 PASS, 0 FAIL, rc=0**.

Caught, for completeness: mask branch inverted (9/3), mask branch always passes (11/1), `else` downgraded from `ck(False)` to `gap()` (10/2). Good.

**F1. A width bug in the mask branch is invisible. This is the sharp one.**

Every fixture in `CASES` uses `8` bits. Every mask field that actually exists is 16, 32 or 64 bits. So clamping the check to the low byte survives:

```python
undef = got[name] & ~m & 0xFF     # the mutation
```

Driven with the six real mask fields from the pinned rev, each given an undefined bit above bit 7:

```
  adp.talker_capabilities     bits=16 mask=0x803F              val=0x100          -> *** ACCEPTED ***
  adp.controller_capabilities bits=32 mask=0x8000              val=0x10000        -> *** ACCEPTED ***
  aecp.acquire_entity_flags   bits=32 mask=0xC0000000          val=0x10000        -> *** ACCEPTED ***
  aecp.lock_entity_flags      bits=32 mask=0x80000000          val=0x10000        -> *** ACCEPTED ***
  aecp.tlv_length_and_mode    bits=16 mask=0xFECF              val=0x10           -> CAUGHT
  aecp.msrp_flags             bits=64 mask=0xF800000000000000  val=0x10000000000  -> *** ACCEPTED ***

  grade_tx self-test: 12 checks: 12 PASS, 0 FAIL       rc=0
```

Five of six real fields wave an undefined bit through, and the gate stays green. That is precisely the failure shape this PR exists to kill: a check that appears to assert something and asserts less. One extra case with a 32-bit field and a high undefined bit closes it. Please add it here.

**F2. The `else` is not proven to name the kind, only the field.**

Deleting `% sorted(con)` from the detail string leaves both needles intact, so it survives:

```
  failure record: (None, 't.f constraint is gradeable', 'ungradeable constraint kind -- teach grade_tx')
  grade_tx self-test: 12 checks: 12 PASS, 0 FAIL       rc=0
```

The kind is gone and the gate does not notice. This is a direct miss against #146's own acceptance list, which says "fails loudly and names the field **and the kind**", and the case is even named `unknown kind names the field`. Adding `"regex"` as a needle on that case is one character of real work.

**F3/F4. Dispatch order is unguarded.** Swapping `range` and `values`, or hoisting `mask` ahead of `value`, both survive at 12/12, because no fixture declares two kinds. Low severity today, and I confirmed no field at the pin declares two either. But tsn-gen documents a precedence (`value`/`values` beats `range` beats `mask`, `docs/mid-level/writing-protocols.md:56`) that nothing in `grade_tx` enforces or tests. Worth a one-line case with `{"value": 5, "mask": [0x0F]}` asserting the value branch wins, if you want the gate to encode the precedence rather than inherit it from source order.

**F5. The self-test has no floor, so it can be emptied silently.** Truncating `CASES` to `[]`:

```
grade_tx self-test: 0 checks: 0 PASS, 0 FAIL         SELFTEST_RC=0
make graderself                                      MAKE_GRADERSELF_RC=0
```

`suite_tally` would call that log NOCOUNT on its own, but the unconditional 2-check `matrix-check` floor rescues the suite, so the sweep goes green:

```
  vacuous alone:  checks=0 failures=0 NOCOUNT?=True
  vacuous+floor:  checks=2 failures=0 NOCOUNT?=False
```

The Makefile's own fifteen-line comment above `aaf` describes this exact hazard and solves it with `suite_tally.py --campaign-guard`. The new target is the one place in this Makefile that does not get that treatment. Either give `graderself` the same tee-and-guard discipline, or assert a minimum case count inside `main()`. Not a blocker, but it is the one thing that would let this gate become a gate that asserts nothing, which is the thing the PR is about.

---

## Verdict

POSITIVE. Merge it. The grader change is correct, the semantics are triple-sourced, the pinned-rev blast radius is provably nil, and the gate does what it says. F1 and F2 are each one line and would close real holes in the new gate; I would like them in this PR. F5 and the missing tracking issue from section 5 are the durability asks, and the tracking issue is the one I would not let slide, since without it #146 closes over a hole nobody owns.

Note for the record: `verilator-suites` and `yosys-portability` were still pending when I reviewed. The other four checks pass, and I reproduced `docs_check.py` locally at `0 finding(s) across 211 md files`. Nothing was committed or pushed; my clone is clean and no existing worktree was touched.

--- a/tb/verilator/tsn_fuzz/fuzz_ptp.py
+++ b/tb/verilator/tsn_fuzz/fuzz_ptp.py
@@ -261,6 +261,11 @@ class Campaign:
         for name, _bits, con in model.fields:
             if name in skip or name not in got or not con:
                 continue
+            # Dispatch order matches tsn_model.legal()/illegal() exactly:
+            # value, values, range, mask. No field declares two kinds today,
+            # but a grader that resolved a combination differently from the
+            # generator that produced the stimulus would be judging by one
+            # rule what was built by another.
             if "value" in con:
                 if name in gaps:
                     self.eq_or_gap("%s.%s" % (label, name), got[name],
@@ -268,14 +273,14 @@ class Campaign:
                 else:
                     self.rep.eq("%s.%s" % (label, name), got[name],
                                 int(con["value"]))
-            elif "range" in con:
-                lo, hi = int(con["range"][0]), int(con["range"][1])
-                self.rep.ck("%s.%s in [%d,%d]" % (label, name, lo, hi),
-                            lo <= got[name] <= hi, "got=%d" % got[name])
             elif "values" in con:
                 self.rep.ck("%s.%s legal" % (label, name),
                             got[name] in [int(v) for v in con["values"]],
                             "got=%d" % got[name])
+            elif "range" in con:
+                lo, hi = int(con["range"][0]), int(con["range"][1])
+                self.rep.ck("%s.%s in [%d,%d]" % (label, name, lo, hi),
+                            lo <= got[name] <= hi, "got=%d" % got[name])
             elif "mask" in con:
                 # A mask names the DEFINED bits; the spec claim is that no bit
                 # outside that set is set. tsn_model already reads masks this

--- a/tb/verilator/tsn_fuzz/test_grade_tx.py
+++ b/tb/verilator/tsn_fuzz/test_grade_tx.py
@@ -78,10 +78,50 @@ CASES = [
      8, {"mask": [0x0F]}, [0x85], 0, 1, "undefined=0x80"),
     ("mask: every defined bit set is still legal",
      8, {"mask": [0x0F]}, [0x0F], 1, 0, ""),
+
+    # REAL WIDTHS. Every mask in the live models is 16, 32 or 64 bits wide
+    # (talker/listener_capabilities 16, controller_capabilities 32,
+    # tlv_length_and_mode 16, acquire/lock_entity_flags 32, msrp_flags 64) and
+    # every undefined bit that matters sits ABOVE bit 7. An 8-bit fixture set
+    # cannot see a grader that truncates -- `got & ~m & 0xFF` passes all three
+    # cases above while waving an undefined bit through five of the six real
+    # fields. Each case below puts the offending bit high enough that any
+    # narrower mask arithmetic misses it.
+    ("mask 16b talker_capabilities: legal",
+     16, {"mask": [0x803F]}, [0x80, 0x3F], 1, 0, ""),
+    ("mask 16b: undefined bit 14, above the low byte",
+     16, {"mask": [0x803F]}, [0x40, 0x00], 0, 1, "undefined=0x4000"),
+    ("mask 32b acquire_entity_flags: legal",
+     32, {"mask": [0xC0000000]}, [0xC0, 0, 0, 0], 1, 0, ""),
+    ("mask 32b: undefined bit 29, above 16 bits",
+     32, {"mask": [0xC0000000]}, [0x20, 0, 0, 0], 0, 1, "undefined=0x20000000"),
+    ("mask 64b msrp_flags: legal",
+     64, {"mask": [0xF800000000000000]},
+     [0xF8, 0, 0, 0, 0, 0, 0, 0], 1, 0, ""),
+    ("mask 64b: undefined bit 58, above 32 bits",
+     64, {"mask": [0xF800000000000000]},
+     [0x04, 0, 0, 0, 0, 0, 0, 0], 0, 1, "undefined=0x400000000000000"),
+
     ("unknown kind fails closed",
      8, {"regex": ["x"]}, [0x00], 0, 1, "ungradeable constraint kind"),
     ("unknown kind names the field",
      8, {"regex": ["x"]}, [0x00], 0, 1, "t.f"),
+    # #146 asks the failure to name the field AND THE KIND. Without this the
+    # message could drop the kind and both needles above would still be met.
+    ("unknown kind names the KIND itself",
+     8, {"regex": ["x"]}, [0x00], 0, 1, "regex"),
+
+    # Dispatch precedence. No field at the pinned rev declares two kinds, so
+    # a reordering of the if/elif chain is otherwise invisible. Pin the order
+    # tsn_model.legal() uses (value, values, range, mask) so the grader and the
+    # generator cannot silently disagree about which constraint governs.
+    ("precedence: value wins over mask",
+     8, {"value": 0x03, "mask": [0x0F]}, [0x01], 0, 1, "exp=3"),
+    ("precedence: values wins over range",
+     8, {"values": [1, 2], "range": [0, 255]}, [9], 0, 1, "legal"),
+    ("precedence: range wins over mask",
+     8, {"range": [1, 9], "mask": [0xFF]}, [200], 0, 1, "in [1,9]"),
+
     ("value: matching", 8, {"value": 0x42}, [0x42], 1, 0, ""),
     ("value: mismatching", 8, {"value": 0x42}, [0x43], 0, 1, "got=67"),
     ("range: inside", 8, {"range": [1, 9]}, [5], 1, 0, ""),
@@ -91,13 +131,22 @@ CASES = [
     # An absent `expected:` block must keep SKIPPING. Failing it here would
     # redden every legitimately unconstrained field (sequence_id, clock
     # identities), so the fail-closed rule deliberately stops at non-empty
-    # constraints. The shrinking-tally problem that causes is tracked apart.
+    # constraints. The shrinking-tally problem that causes is issue #150.
     ("no constraint at all still skips",
      8, {}, [0xFF], 0, 0, ""),
 ]
 
+# A floor, for the same reason the campaigns carry one: an empty CASES list
+# would print "0 checks: 0 PASS, 0 FAIL" and exit 0, which reads as success.
+MIN_CASES = 22
+
+
 
 def main():
+    if len(CASES) < MIN_CASES:
+        print("grade_tx self-test: FAIL, only %d cases, expected at least %d"
+              % (len(CASES), MIN_CASES))
+        return 1
     bad = 0
     for name, bits, con, pdu, want_p, want_f, needle in CASES:
         npass, nfail, msg = grade(bits, con, pdu)

